node/src/protocol/presignature.rs

```diff
@@ -0,0 +1,228 @@
+    /// 1) Already generated in which case returns `None`, or
+    /// 2) Is currently being generated by `protocol` in which case returns `Some(protocol)`, or
+    /// 3) Has never been seen by the manager in which case start a new protocol and returns `Some(protocol)`, or
+    /// 4) Depends on triples (`triple0`/`triple1`) that are unknown to the node
+    // TODO: What if the presignature completed generation and is already spent?
+    pub fn get_or_generate(
+        &mut self,
+        id: PresignatureId,
+        triple0: TripleId,
+        triple1: TripleId,
+        triple_manager: &mut TripleManager,
+        public_key: &PublicKey,
+        private_share: &PrivateKeyShare,
+    ) -> Option<&mut PresignatureProtocol> {
+        if self.presignatures.contains_key(&id) {
+            None
+        } else {
+            match self.generators.entry(id) {
+                Entry::Vacant(entry) => {
+                    tracing::info!(id, "joining protocol to generate a new presignature");
+                    let triple0 = match triple_manager.take(triple0) {
+                        Some(triple0) => triple0,
+                        None => {
+                            tracing::warn!(triple_id = triple0, "triple0 is missing, can't join");
+                            return None;
+                        }
+                    };
+                    let triple1 = match triple_manager.take(triple1) {
+                        Some(triple1) => triple1,
+                        None => {
+                            tracing::warn!(triple_id = triple1, "triple1 is missing, can't join");
+                            return None;
+                        }
+                    };
+                    let protocol = Box::new(
+                        cait_sith::presign(
+                            &self.participants,
+                            self.me,
+                            PresignArguments {
+                                triple0: (triple0.share, triple0.public),
+                                triple1: (triple1.share, triple1.public),
+                                keygen_out: KeygenOutput {
+                                    private_share: *private_share,
+                                    public_key: *public_key,
+                                },
+                                threshold: self.threshold,
+                            },
+                        )
+                        .unwrap(),
+                    );
+                    Some(
+                        &mut entry
+                            .insert(PresignatureGenerator {
+                                protocol,
+                                triple0: triple0.id,
+                                triple1: triple1.id,
+                            })
+                            .protocol,
+                    )
+                }
+                Entry::Occupied(entry) => Some(&mut entry.into_mut().protocol),
```

Comment:
Seems like we can just reuse the `generate` function instead of rewriting it here